attach/cid.c

```diff
@@ -0,0 +1,235 @@
+/**
+ * @file
+ * Attachment Content-ID header functions
+ *
+ * @authors
+ * Copyright (C) 2022 David Purton <dcpurton@marshwiggle.net>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include "email/lib.h"
+#include "cid.h"
+#include "attach.h"
+#include "mailcap.h"
+#include "mutt_attach.h"
+#include "muttlib.h"
+
+/**
+ * cid_map_free - Free a CidMap
+ * @param[out] ptr CidMap to free
+ */
+void cid_map_free(struct CidMap **ptr)
+{
+  if (!ptr)
+    return;
+
+  struct CidMap *cid_map = *ptr;
+
+  FREE(&cid_map->cid);
+  FREE(&cid_map->fname);
+
+  FREE(ptr);
+}
+
+/**
+ * cid_map_new - Initialise a new CidMap
+ * @retval ptr Newly allocated CidMap
+ */
+struct CidMap *cid_map_new(const char *cid, const char *filename)
+{
+  if (!cid || !filename)
+    return NULL;
+
+  struct CidMap *cid_map = mutt_mem_calloc(1, sizeof(struct CidMap));
+
+  cid_map->cid = mutt_str_dup(cid);
+  cid_map->fname = mutt_str_dup(filename);
+
+  return cid_map;
+}
+
+/**
+ * cid_save_attachment - Save attachment if it has a Content-ID
+ * @param[in]  body         Body to check and save
+ * @param[out] cid_map_list List of Content-ID to filename mappings
+ *
+ * If body has a Content-ID, it is saved to disk and a new Content-ID to filename
+ * mapping is added to cid_map_list.
+ */
+static void cid_save_attachment(struct Body *body, struct CidMapList *cid_map_list)
+{
+  if (!body || !cid_map_list)
+    return;
+
+  char *id = mutt_param_get(&body->parameter, "content-id");
+  struct Buffer *tmpfile = mutt_buffer_pool_get();
+  struct Buffer *cid = mutt_buffer_pool_get();
+  bool has_tempfile = false;
+  FILE *fp = NULL;
+
+  if (id)
```

Comment:
If `!id`, then we don't do anything, so we could move this line above the two `mutt_buffer_pool_get()` lines.
```c
char *id = mutt_param_get(&body->parameter, "content-id");
if (!id)
  return;
```

That'd simplify the error path.

Reply:
I left the `goto` in, but could clean up and return if `mutt_save_attachment()` fails if you think better.

Reply:
The `goto`'s fine